core/templates/dev/head/domain/collection/CollectionRightsObjectFactory.js

```diff
@@ -0,0 +1,113 @@
+      this._ownerNames = collectionRightsObject.owner_names;
+    };
+
+    // Instance methods
+
+    CollectionRights.prototype.getCollectionId = function() {
+      return this._collectionId;
+    };
+
+    // Returns true if the the user can edit the collection. This property is
+    // immutable.
+    CollectionRights.prototype.canEdit = function() {
+      return this._canEdit;
+    };
+
+    // Returns true if the user can edit the collection. This property is
+    // immutable.
+    CollectionRights.prototype.canUnpublish = function() {
+      return this._canUnpublish;
+    };
+
+    // Returns true if the collection is private.
+    CollectionRights.prototype.isPrivate = function() {
+      return $rootScope.isPrivate;
+    };
+
+    // Sets isPrivate to false only if the
+    // collectionRightsBackendObject.is_private passed in is false; otherwise,
+    // writes an error to the log. Assumes that the backend returns a success
+    // in set public operation. Returns the result of isPrivate regardless of
+    // error.
+    CollectionRights.prototype.setPublic = function(
+      collectionRightsBackendObject) {
+      if (!collectionRightsBackendObject.is_private) {
+        $rootScope.isPrivate = false;
+      } else {
+        $log.error(
+          'Backend indicated a collection was successfully ' +
+          'published, but response.data.is_private returned true.');
+      }
+      return this.isPrivate();
+    };
+
+    // Sets isPrivate to true only if the
+    // collectionRightsBackendObject.is_private passed in is true AND
+    // canUnpublish is true. If canUnpublish is true and is_private is false,
+    // it writes an error to the log. Assumes that the backend returns a
+    // success in set private operation. Returns the result of isPrivate
+    // regardless of error.
+    CollectionRights.prototype.setPrivate = function(
+      collectionRightsBackendObject) {
+      if (this.canUnpublish()) {
+        if (collectionRightsBackendObject.is_private) {
+          $rootScope.isPrivate = true;
+        } else {
+          $log.error(
+            'Backend indicated a collection was successfully ' +
+            'unpublished, but response.data.is_private returned false.');
+        }
+        return this.isPrivate();
+      }
+    };
+
+    // Returns the owner names of the collection. This property is immutable.
+    CollectionRights.prototype.getOwnerNames = function() {
+      return this._ownerNames;
+    };
+
+    // Static class methods. Note that "this" is not available in static
+    // contexts. This function takes a JSON object which represents a backend
+    // collection python dict.
+    CollectionRights.create = function(collectionRightsBackendObject) {
+      return new CollectionRights(collectionRightsBackendObject);
+    };
+
+    // Create a new, empty collection rights. This is not guaranteed to pass
+    // validation tests.
+    CollectionRights.createEmptyCollectionRights = function() {
```

Comment:
This is interesting -- my first thought was "why do we need to do this?" and then I realized that you need it for what happens before the AJAX response. My issue with it is that it sweeps the undefined behavior under the rug a bit -- it's not clear that "undefined" is an important possibility for the value of some of these fields, and that it actually means something.

Perhaps a way to mitigate this is to rename this to createDummyCollectionRights, and have the docstring explain that the returned object is intended as a placeholder until the actual rights data comes from the backend. Also, in the constructor, document the meanings of each of the fields, and explain in particular what "undefined" means for each of them. That way, we have a canonical reference.

What do you think?